### PR TITLE
Best Checkpointing Hook

### PR DESCRIPTION
**Pull request by levan92**
Implemented hook for saving best model, as discussed in #2114

### CODE REVIEW

**Comment by levan92** [comment]
Hmm sorry, not sure why there is a test failure, issue seems to stem from `detectron2.data.build.DatasetFromList` but that wasn't changed in this PR

File: detectron2/engine/hooks.py [hunk 1]
```diff
@@ -199,6 +201,81 @@ def after_step(self):
         self.step(self.trainer.iter)
 
 
+class BestCheckpointer(HookBase):
+    """
+    Checkpoints best weights based off given metric.
+
+    This hook should be used in conjunction to and executed after `EvalHook`.
```
Comment (ppwwyyxx) [inline detectron2/engine/hooks.py:208]:
```suggestion
    This hook should be used in conjunction to and executed after the hook that produces the metric, e.g. `EvalHook`.
```

File: detectron2/engine/hooks.py [hunk 2]
```diff
@@ -199,6 +201,81 @@ def after_step(self):
...
+        file_prefix: str = "model_best",
+        append_iter: bool = False,
+    ) -> None:
+        """
+        Args:
+            eval_period (int): the period `EvalHook` is set to run.
+            checkpointer: the checkpointer object used to save checkpoints.
+            val_metric (str): chosen validation metric to track for best checkpointing
+            file_prefix (str): the prefix of checkpoint's filename, defaults to "model_best"
+            append_iter (bool): if True, checkpoint file name will have iteration num appended, else write to the same checkpoint file.
```
Comment (ppwwyyxx) [inline detectron2/engine/hooks.py:225]: There is no harm to save iter (it's saved in all other checkpoints anyway). Let's use True and remove this option.
However I don't think we should add iter to the saved file name - it will then save __many__ best models instead of one.

File: detectron2/config/defaults.py [hunk 3]
```diff
@@ -531,6 +531,10 @@
 # Save a checkpoint after every this number of iterations
 _C.SOLVER.CHECKPOINT_PERIOD = 5000
 
+# Used with `hooks.BestCheckpointer`. Checkpoints every best model based on given eval metric.
+_C.SOLVER.BEST_CHECKPOINTER = CN({"ENABLED": False})
+_C.SOLVER.BEST_CHECKPOINTER.METRIC = "bbox/AP50"
+
```
Comment (ppwwyyxx) [inline detectron2/config/defaults.py:537]: please remove new config options as we expect users to add the hook to their training script directly

File: detectron2/engine/hooks.py [hunk 4]
```diff
@@ -199,6 +201,81 @@ def after_step(self):
...
+                additional_state = {"iteration": metric_iter}
+                save_name = f"{self._file_prefix}_{metric_iter:07d}"
+            else:
+                save_name = f"{self._file_prefix}"
+                additional_state = {}
+            self._checkpointer.save(save_name, **additional_state)
+            if self.best_metric is None:
+                self._logger.info(
+                    f"Saved first model with latest eval score for {self._val_metric} at {latest_metric:0.5f}"
+                )
```
Comment (ppwwyyxx) [inline detectron2/engine/hooks.py:249]: Saving the first model looks useless. Could you change the behavior to not save?

File: detectron2/engine/hooks.py [hunk 5]
```diff
@@ -199,6 +201,81 @@ def after_step(self):
...
+    Checkpoints best weights based off given metric.
+
+    This hook should be used in conjunction to and executed after `EvalHook`.
+    """
+
+    def __init__(
+        self,
+        eval_period: int,
+        checkpointer: Checkpointer,
+        val_metric: str = "bbox/AP50",
```
Comment (ppwwyyxx) [inline detectron2/engine/hooks.py:215]:
```suggestion
        val_metric: str,
```

File: detectron2/engine/hooks.py [hunk 6]
```diff
@@ -199,6 +201,81 @@ def after_step(self):
...
+        file_prefix: str = "model_best",
+        append_iter: bool = False,
+    ) -> None:
+        """
+        Args:
+            eval_period (int): the period `EvalHook` is set to run.
+            checkpointer: the checkpointer object used to save checkpoints.
+            val_metric (str): chosen validation metric to track for best checkpointing
```
Comment (ppwwyyxx) [inline detectron2/engine/hooks.py:223]:
```suggestion
            val_metric (str): chosen validation metric to track for best checkpoint, e.g. "bbox/AP50"
```

File: detectron2/engine/hooks.py [hunk 7]
```diff
@@ -199,6 +201,81 @@ def after_step(self):
...
 
 
+class BestCheckpointer(HookBase):
+    """
+    Checkpoints best weights based off given metric.
+
+    This hook should be used in conjunction to and executed after `EvalHook`.
+    """
+
+    def __init__(
```
Comment (ppwwyyxx) [inline detectron2/engine/hooks.py:211]: please add an option `mode="max"` similar to the one in https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/ModelCheckpoint, with two possible values "min" and "max"

File: detectron2/engine/hooks.py [hunk 8]
```diff
@@ -199,6 +201,81 @@ def after_step(self):
...
+            if self.best_metric is None:
+                self._logger.info(
+                    f"Saved first model with latest eval score for {self._val_metric} at {latest_metric:0.5f}"
+                )
+            else:
+                self._logger.info(
+                    f"Saved best model as latest eval score for {self._val_metric} at {latest_metric:0.5f} is better than last best score at {self.best_metric:0.5f} @ {self.best_iter} steps"
+                )
+            if math.isnan(latest_metric):
+                latest_metric = -1.0
```
Comment (ppwwyyxx) [inline detectron2/engine/hooks.py:255]: a metric can be negative (e.g. log likelihood). Please handle NaNs/Infs more properly: check and never treat them as the best

**Review by ppwwyyxx — COMMENTED** [review]
Thanks! looks pretty reasonable overall and please disregard the test failure about `DatasetFromList`

File: detectron2/engine/hooks.py (hunk 7 above)
Comment (levan92) [reply detectron2/engine/hooks.py:213]: oh right, sorry! totally skipped my mind that the metric could go the other direction :facepalm:

File: detectron2/engine/hooks.py (hunk 4 above)
Comment (levan92) [reply detectron2/engine/hooks.py]: I've modified it such that it only updates the `best_metric` during the "first model" but does not save it. 

However, I do have some lingering concerns for this -- should we just save it as long as the current metric value is valid (not NaN nor infs), even if it is the first model? Seems like it is implemented this way in Keras as well, as seen [here](https://github.com/keras-team/keras/blob/f43c792c129ffba27943c00f11c73c7d8db2838f/keras/callbacks.py#L1412). It initialises the best variable as either negative or positive infinity, then compares with current value and saves even if it is the "first call". We could implement that in a similar fashion here.

Comment (Kevin-Delnoije) [reply detectron2/engine/hooks.py]: Would it be a good idea to add a starting threshold as a keyword argument defaulted to None?
Often when I train models on my dataset I know what to expect from the metric and if it performs a lot worse then I don't want to save it.

Comment (levan92) [reply detectron2/engine/hooks.py]: @ppwwyyxx I do realise that sometimes saving the model after the first evaluation is needed. It is always possible the best model is reached by the first evaluation run, albeit not ideal, and especially if the first evaluation happens after many steps. 

I will pull this commit into this PR if you feel it is alright: https://github.com/levan92/detectron2/commit/59bebb201f59519e8db80f81c0ad2f06347053e0 It exposes a `save_first` argument which default to `False`.

Comment (ppwwyyxx) [reply detectron2/engine/hooks.py]: That argument makes total sense to me. We can let it be the default behavior (i.e. also save the first evaluation).

**Comment by ppwwyyxx** [comment]
Could you add a unittest in `tests/test_engine.py` for the hook? 

In the unittest you can fake an evaluate function like what's done in `test_eval_hook()` in `tests/test_engine.py`, e..g:
```python
metrics = iter([0.3, 0.4, 0.35, 0.5])
hooks.EvalHook(20, lambda: {"metrics": next(metrics)})
```

**Comment by levan92** [comment]
> Sorry for the delay. Could you add a unittest in `tests/test_engine.py` for the hook?
> 
> In the unittest you can fake an evaluate function like what's done in `test_eval_hook()` in `tests/test_engine.py`, e..g:
> 
> ```python
> metrics = iter([0.3, 0.4, 0.35, 0.5])
> hooks.EvalHook(20, lambda: {"metrics": next(metrics)})
> ```

Added in 935939b

**Review by ppwwyyxx — APPROVED** [review]
Thanks!

**Comment by levan92** [comment]
Sorry @ppwwyyxx, found a minor bug which throws when the eval score is nan in the first evaluation. Just pushed the fix for it.

**Comment by AKMourato** [comment]
Hi @levan92, can you give an example of how to use it with the new baselines configs?
Just adding:
```
hooks.BestCheckpointer(test_period, checkpointer, metric_name, mode=mode, save_first=save_first)])
```
is not sufficient I guess?
Thanks

**Comment by levan92** [comment]
Hi @AKMourato, it should be sufficient if you're hardcoding the arguments/passing them in through other ways. 

However, if you want to control these through config files, you will need to do the additional work of adding them into your configs. I've included below an example of how I do it with the normal config style (yaml). But if you're talking about the new_baselines (which uses the "lazy" config style), it should be even easier, just that you may need to define your own `train` config dict (see `configs/common/train.py`) and modify `lazyconfig_train_net.py` accordingly.

Also, if you're doing multi-gpu training, you might want to condition it with a `if comm.is_main_process()`. 

In `train_net.py`:
```python3
def add_custom_configs(cfg: CfgNode):
    _C = cfg
    _C.SOLVER.BEST_CHECKPOINTER = CfgNode({"ENABLED": False})
    _C.SOLVER.BEST_CHECKPOINTER.METRIC = "bbox/AP50"
    _C.SOLVER.BEST_CHECKPOINTER.MODE = "max"
```

```python3
def setup(args):
    """
    Create configs and perform basic setups.
    """
    cfg = get_cfg()
    add_custom_configs(cfg)
    cfg.merge_from_file(args.config_file)
    cfg.merge_from_list(args.opts)
    cfg.freeze()
    default_setup(cfg, args)
    return cfg
```

Copy and override the `build_hooks` method of `DefaultTrainer` to include this after the `EvalHook` [line](https://github.com/facebookresearch/detectron2/blob/55c0078b1ffa6fca21834db7b3cf85c48d66109d/detectron2/engine/defaults.py#L457)
```python3
if cfg.SOLVER.BEST_CHECKPOINTER and comm.is_main_process():
    ret.append(hooks.BestCheckpointer(cfg.TEST.EVAL_PERIOD, self.checkpointer, cfg.SOLVER.BEST_CHECKPOINTER.METRIC, mode=cfg.SOLVER.BEST_CHECKPOINTER.MODE, save_first=True))
```

**Comment by ZahraAnam** [comment]
Hi! @levan92 I tried this code from yours but it didn't save any model as best, did it work for you?

**Comment by sjdch** [comment]
我是小白，想保存最佳的ap,ap50,ap75等信息，应该修改哪些文件，如何修改？谢谢您的回复
